### PR TITLE
Merge values for provider group in the constructor

### PR DESCRIPTION
**Pull request by alsamylkin**
ProviderGroup can and should merge values on construction, instead of on value retrieval.
To do that we need need apply expand to all nested YAML nodes, instead of only top level ones.

### CODE REVIEW

**Comment by glibsm** [comment]
What about `p.Get("foo").Value()` now that we're driving people to the casting libraries? Sounds like they will just get the default value, rather than exposing the error

**Comment by alsamylkin** [comment]
Yeah, there are some other corner cases while populating values inside,  e.g. there is a merge error for populating a first element in the `[]interface{}` slice. Need to think more about it.

File: provider_group.go [hunk 1]
```diff
@@ -21,56 +21,44 @@
...
 // A and B and values from B will overwrite values of A. e.g.
 //   A:                B:                 merge(A, B):
 //     keep:A            new:B              keep:A
 //     update:fromA      update:fromB       update:fromB
 //                                          new:B
 //
-// * if A is a map and B is not, this function will panic,
-// e.g. key:value and -slice
+// If A is a map and B is not, this function will return a Value with
+// an error inside.
```
Comment (glibsm) [inline provider_group.go]: Is there a case when there is a value with an error inside that will then propagate it to `Populate()` and `Value()` method? Or this comment now obsolete

Comment (alsamylkin) [inline provider_group.go]: It can be an error value from the provider, not a merge error. We construct only correct providers now, no runtime surprises.

**Comment by breerly** [comment]
@alsamylkin Is this WIP or ready for review?

**Comment by alsamylkin** [comment]
It is ready for review.

File: yaml.go [hunk 2]
```diff
@@ -82,7 +82,8 @@ func newYAMLProviderCore(files ...io.ReadCloser) (*yamlConfigProvider, error) {
 //     update:fromA      update:fromB       update:fromB
 //                                          new:B
 //
-// * if A is a map and B is not, this function will panic, e.g. key:value and -slice
+// * if A is a map and B is not, this function will return an error,
```
Comment (amckinney) [inline yaml.go]: For consistency, consider updating `* if` to `If` same as in `* in` below.

File: value.go [hunk 3]
```diff
@@ -71,7 +71,11 @@ func (cv Value) WithDefault(value interface{}) (Value, error) {
 		return cv, err
 	}
 
-	g := NewProviderGroup("withDefault", p, cv.provider)
+	g, err := NewProviderGroup("withDefault", p, cv.provider)
+  if err != nil {
```
Comment (glibsm) [inline value.go]: formatting is odd here

Comment (alsamylkin) [inline value.go]: This is an only merge tool.

**Pull request merged**